File: 2020/11/week3/minhyungPark/algorithm/LeetCode1646.kt
/**
 * Leetcode 1646. Get Maximum in Generated Array
 * https://leetcode.com/contest/weekly-contest-214/problems/get-maximum-in-generated-array/
 */

class LeetCode1646 {

    fun getMaximumGenerated(n: Int): Int {
        val nums = IntArray(n+1)
        for (i in 2 until nums.size) {
            if (i%2==0) {
                nums[i] = nums[i/2]
            } else {
                nums[i] = nums[i/2] + nums[i/2 + 1]
            }
        }

        return nums.max()!!
    }
}

File: 2020/11/week4/1ilsang/algorithm/leetcode1646.js
/**
 * https://leetcode.com/contest/weekly-contest-214/problems/get-maximum-in-generated-array/
 * 1646. Get Maximum in Generated Array
 * @param {number} n
 * @return {number}
 */
const getMaximumGenerated = (n) => {
  if (n < 2) return n;
  let answer = 1;
  const arr = [0, 1];
  for (let i = 2; i <= n; i++) {
    const quotient = Math.floor(i / 2);
    const value = i % 2 ? arr[quotient] + arr[quotient + 1] : arr[quotient];
    if (answer < value) answer = value;
    arr[i] = value;
  }
  return answer;
};
